### ISSUE TITLE
Table Filters: : sorting feature is included in the TH, thus unnecessarily referenced by screen readers when navigating the table 

### ISSUE THREAD

**Issue by StacyBleeks** [wet-boew/wet-boew#9127]
Within a table that offers sorting functionality, the sorting functionality is included in the relevant column header. so, "sort column alphabetically" is in the column header. this is fine, but for screen reader users, every time the column header is referenced, the sort functionality should not be referenced. In other words, the sorting functionality should be offered in the column header data cell, but should not be referenced as part of the TH.
Here is a WET working example of where this issue occurs: https://wet-boew.github.io/v4.0-ci/demos/tables/tables-fltrs-en.html#dataset-filter2
And here is a Deque example of how to make it work so that the sort functionality is not referenced every time a screen reader navigates to the sortable column within the table.https://dequeuniversity.com/library/aria/table-sortable
https://dequeuniversity.com/library/aria/table-sortable

**Comment by StacyBleeks** [comment]
@thekodester  Do you have any thoughts about this issue? Let me know if I should clarify further, or if a demo would be helpful.

**Comment by thekodester** [comment]
If we break down the code of the two examples they both use the header tag (`th`).

WET currently uses span tags to create the sorting icons where as the other example uses buttons. This is added by WET.

The 3rd party plugin DataTables does not (https://datatables.net):
`<th class="sorting" tabindex="0" aria-controls="example" rowspan="1" colspan="1" style="width: 120px;" aria-label="Name: activate to sort column ascending">Name</th>`

https://wet-boew.github.io/v4.0-ci/demos/tables/tables-fltrs-en.html#dataset-filter2
```
<th class="sorting" tabindex="0" aria-controls="dataset-filter2" rowspan="1" colspan="1" aria-label="First Name: activate for ascending sort" style="width: 192px;">
First Name
<span class="sorting-cnt"><span class="sorting-icons"></span></span>
</th>
```

https://dequeuniversity.com/library/aria/table-sortable
```
<th role="columnheader" scope="col">
<button class="sortableColumnLabel" tabindex="0">
"Name"
</button>
</th>
```

Can you see if you reproduce the issue with default code from DataTable?

**Comment by StacyBleeks** [comment]
@thekodester  Which one of the links you provided would you like me to check?

**Comment by StacyBleeks** [comment]
@thekodester  would you be able to meet for 20 minutes about this issue sometime in July? can we meet via Teams? I feel a breif chat would help in this case. I think this issue would benefit from a live chat. Thanks!

**Comment by thekodester** [comment]
Would love too, but my schedule does not really allow for it. Recently took on overnight supervisor.

If I find the time I might be able to create demo page with multiple approaches that could be used to test which approach works.

**Comment by StacyBleeks** [comment]
@thekodester  if you could create a few test examples, I would be happy to test them. 
I am sure you've seen the examples online where the table sorting feature within the TH is simply a button, where only one instance of the sorting feature is visible at a time. 
while this is an interesting approach, the ideal situation would effectively not reference the sorting info for screen readers when navigating within a table. In other words, the sorting feature is relevant within the table cell when the focus is actually on that cell, but when the TH is being referenced, the sorting fuctionality within the tH is no longer relevant.

If you can think of some wizardry that could resolve this issue, it would be awesome! Seriusly this issue has been a puzzle for quite some time.
I was once told that there would be no way to create "accessible footnotes, but WEt made that dream real!
In any case, if you build it, I will test it :)

**Comment by StacyBleeks** [comment]
@thekodester  check out this Adrian Roselli working example posted in April 2021 https://adrianroselli.com/2021/04/sortable-table-columns.html
I would love to demo it for you using a screen reader with sound on via Teams. It would only take 5 minutes. By no means do I think the Adrian Roselli ARIA hidden based example is perfect, but I believe it is quite good.

**Comment by StacyBleeks** [comment]
@thekodester  I can see how this is a low importance issue. So, if you have time to take a crack at it, I'm happy to test it whenever. Cheers! 😀

**Comment by thekodester** [comment]
Apologizes I am just a volunteer on this GitHub project currently, so not as quick as I use to be (lower on my list).

Regardless I tired to take a look at what we can do only I was able to do was remove the spans added by wet and use CSS icons instead, unsure if that well affect or achieve your goal. I'll see if I can get a demo up and running.

As for removing the aria label generated by DataTables I could not find an easy way to remove or update this text (`aria-label="Name: activate to sort column descending"`) as it is added by the 3rd party plugin I could not locate any way to modify it. I searched a quite a few of their articles on https://datatables.net.

**Comment by thekodester** [comment]
Here demo with span sorting icons removed: https://thekodester.ca/wet/demos/tables/sorting/css-icons

**Comment by StacyBleeks** [comment]
@thekodester  thanks for working on this. I know from the various online discussions about this issue, the solution is not easy.

Regarding the demo table: Screen readers are still referencing "activate for ascending sort" for every column header when navigating across a row within the table . I believe that text is only valuable when on the TH.

**Comment by thekodester** [comment]
Yes DataTables adds that text by default to the TH tags.

**Comment by thekodester** [comment]
Regardless let me know if the change with using CSS icons for sorting makes a different.

https://thekodester.ca/wet/demos/tables/sorting/css-icons

**Comment by StacyBleeks** [comment]
@thekodester  viewing https://thekodester.ca/wet/demos/tables/sorting/css-icons the sorting feature is only announced by screen readers for one column.
When navigating the table the TH label is referenced when in any and every cell "activate for ascending sort". So, this makes it seem like the cell can be "activated for ascending sort". The constant repetitious announcement when navigating from cell to cell adds confusion.

The column header TH doesn't seem to have a button so blind screen reader users wil not know that there is a sort feature availabel on the column headers TH.

Questions for you:
1. Can you remove the "sort for ascending sort "from the TH label (or is anything in the column header going to be part of the tH?
2. 2. Would it make sense to include a note in the table summary, that the top row column headers offer a sorting functionality. I know the table summary is depricated but this would be good info for screen reader users so that they know the sorting functionality is available. TH label e

**Comment by thekodester** [comment]
I can possibly give it a try this week sometime.

Alternative maybe to the summary is maybe rewording the “sort for ascending” to “press to sort column name ascending”?

Like is the column was called dates. The aria tag be “press to sort Dates ascending”?

**Comment by StacyBleeks** [comment]
@thekodester  that's a good idea. What about the issue of the sorting info being referenced by screen readers for every cell within the table. 
is it possible to have text within the column header cell but not within the TH label (so that info is only identified/announced by screen readers when on the column header cell?

Alternatively I have seen a working example where the column headers are sortable and include a button. The sorting info is identified visually with an up/down arrrow. If there was info in a table summary (or ARIA label) to say that the column header buttons toggle ascending sorting.
I do like the idea of an aria tag be “press to sort Dates ascending” if this info is not referenced when navigating within table cells, rather only when in the column headers.

**Comment by thekodester** [comment]
Unable to do anything atm the since I cannot get the project to setup, but I have submitted an issue tracker https://github.com/wet-boew/wet-boew/issues/9155 to see if it is my system or the project.

**Comment by thekodester** [comment]
@StacyBleeks Here a new demo https://thekodester.ca/wet/demos/tables/sorting/button

**Comment by thekodester** [comment]
Any update?

**Comment by StacyBleeks** [comment]
@thekodester  This design is  excellent.! The column header's sorting text is no longer being referenced by screen readers when navigating through every cell in the table, and the sorting feature still works well for each column.
Since only one instance seems to be identified and the non active sorting columns are currently unlabelled buttons,  would you be able to add some hidden text in the table caption to indicate that this table provides a sorting feature via the buttons across the column header row with only one instance visible at a time   

This would be a good case to use the table summary. I believe the table summary attribute is depricated but it would be good to let blind screen readers know that the buttons across the top row will apply the sorting filter on their respective columns.

What do you think?the table summary is

**Comment by thekodester** [comment]
Caption would not be simple. Already issues around way datatables loads them. As for summary tag should be simple to add. i18n text database would need to be updated. I'll create a PR and update the demo.

**Comment by thekodester** [comment]
Update demo with summary tag. Also on VoiceOver it does indicate what the sort button does. Unsure about other screenreaders.

<img width="681" alt="Screen Shot 2021-09-28 at 4 04 34 PM" src="https://user-images.githubusercontent.com/23484460/135157973-d8cf3bb5-cf70-4414-ab99-6e1877f055ed.png">

**Comment by thekodester** [comment]
PR is submitted also.

**Comment by StacyBleeks** [comment]
@thekodester  I have tested the demo page using IOS 14 with Voiceover enabled. There are a few observations I'd like to share with you:
 With VO/iphone testing using IOS14  
1. the first instace of the sorting feature is announced on the first button that appears in the top header row; something like user agent button sort up".
2. When actionning this button using a single finger double tap, the "sort up" text is gone and does not appear on any of the other unlabelled buttons across the top row. In fact actioning any of the buttons at this point doesn't cause the "sort up" text to appear on the actioned button (like it does with the desktop based Chrome browser.
3. The summary is very useful for low vision/blind screen reader users and while it does work on desktop browsers, the table summary is nnot idenntified by VO.
4. Actionning any of the sorting buttons does perform the sorting functionality for that column but there is little context provided to VO users as to the sorting functionality that it's almost guesswork since each button has no label.
Let me know if I can clarify any of this for you. 
Thanks!

**Comment by thekodester** [comment]
> @thekodester I have tested the demo page using IOS 14 with Voiceover enabled. There are a few observations I'd like to share with you: With VO/iphone testing using IOS14
> 
> 1. the first instace of the sorting feature is announced on the first button that appears in the top header row; something like user agent button sort up".
> 2. When actionning this button using a single finger double tap, the "sort up" text is gone and does not appear on any of the other unlabelled buttons across the top row. In fact actioning any of the buttons at this point doesn't cause the "sort up" text to appear on the actioned button (like it does with the desktop based Chrome browser.

Issue 1 and 2 are basically the same thing. Live region is being triggered probably aria-sort triggering it.

The live region well stop triggering on each column once a single column is actively slowly sorting. This seems like common behaviour on other sites shared here.

> 3. The summary is very useful for low vision/blind screen reader users and while it does work on desktop browsers, the table summary is nnot idenntified by VO.

Limit of voiceover.

> 4. Actionning any of the sorting buttons does perform the sorting functionality for that column but there is little context provided to VO users as to the sorting functionality that it's almost guesswork since each button has no label.

Adding a label well reintroduced original issue. For some reason all screen readers treat any aria-label or label tag as an identifier for all cells.

Don’t think span with invisible text would work as screen readers use the button text also as an identifier for each cell. I can test to confirm.
 
>    Let me know if I can clarify any of this for you.
>    Thanks!

**Comment by StacyBleeks** [comment]
@thekodester  I think the new design addresses the issue of the sorting info, within the column header, being referenced by screen readers when navigating each and every cell. That behaviour causes confusion since the only the column header content should be referenced, onot the column header content + the sorting feature info.

The recent sorting table demo is less verbose for sure but I am not confident that blind/low vision screen reader users will understand that each of the buttons across the top row will cause the cells within the column to "sort".

I like the new design, but I have the advantage of already knowing that the top row buttons will sort the columns. 
It sucks that the table summary is depricated. 
Could you add some hidden text to a table caption that offers the same info used in the table summary? I'd really like to test that. I know the table caption is not manditory for all tables but it is a "best practise" and in this case, would offer a conduit for informing pescreen reader users about the sorting feature.

BTW- I really appreciate the work you have put into this. for every cell within

**Comment by thekodester** [comment]
Someone we'll have to take over. I cannot put anymore time into this.

###### I am just a volunteer on this GitHub project.
